### ISSUE TITLE
CIBox releases and update path. LTS talk

### ISSUE THREAD

**Issue by podarok** [propeoplemd/cibox#436]
Hello, dear team

The problems
- we've got a plenty instances of #cibox and all they are old versions. 
- development and refactoring of cibox is much more faster then instance's life. Updating almost impossible at current state of our project
- we have growing number of forks (35 at the moment) and backporting from them is a pain
- we are and will be getting a lot of requests from the old instances about fixing bugs that already fixed in a new version but it is a pain to incorporate fixes back to the old instances
- when the "speed of development is much more preferred than security, code quality and team grow and risks"  we should be able to provide a solution for how to beat this rule without revolution in a company

So here is a proposal
- Make all software in current cibox sticked to specific versions, even when it is not obvious (like we have dependencies for apache, but they are different every time cibox installed)
- Create CIBox LTS version after getting everything stable.
- LTS versions should have update path (commercial support as well)
- LTS versions should have a rethinked recipe for how to work with an enterprise projects like @propeoplemd/cypress-pp where we have a lot of servers for a production and getting builds is a pain. We have lack of tools for minimizing production dumps and optimizing reinstall process
- LTS versions should have an ability to disable all reinstall steps by users. We should have features.yml config for that and good documentation.

I'd appreciate any input to this

**Comment by asgorobets** [comment]
> - Make all software in current cibox sticked to specific versions, even when it is not obvious (like we have dependencies for apache, but they are different every time cibox installed)

Yes, this is what we do for Drupal projects built with Drush Make or Composer, it's way safer to keep fixed versions rather than rely on package manager to grab an untested version automatically and breaking the build or the box. But if we do this we'll need to have someone, or some process of identifying what versions have security releases and need to be updated in next bugfix release cycle.

> - Create CIBox LTS version after getting everything stable.
> - LTS versions should have update path (commercial support as well)

Yes, yes and yes, and LTS should be basically every MAJOR release as we already use semantic versioning. Refer [Semantic Versioning](http://semver.org/)
```
MAJOR version when you make incompatible API changes,
MINOR version when you add functionality in a backwards-compatible manner, and
PATCH version when you make backwards-compatible bug fixes.
```

I propose we have a non-breaking path for all minor releases, if some new features is added that requires some files to be moved around or refactoring to be done, we need an upgrade path for this change, unless we do major refactoring and switch to MAJOR version, in this case we can skip the upgrade path. Drupal does that for years, we should also do, think hook updates ;)
Project repository and CIBOX repository are different repositories, it is not possible to merge their history and to do anything on Git level other than just dropping all files over and hoping that nothing was removed during the latest release cycle. And for this reason I think it would be more reasonable to have a script that updates your project repository files related to CIBOX with latest version on demand. Something similar like [Drupal Composer](https://github.com/drupal-composer/drupal-project) does with drupal-scaffolding project (composer plugin). Drupal Composer promotes you to update your core using composer update drupal/core, but everything which is not part of drupal/core project in Composer (non-source files like .htaccess, .gitignore, sites/default...) is downloaded separately by downloading a fresh Drupal of specified version and preparing a list of files form the download to be applied on top of current repository. Refer [this](https://github.com/drupal-composer/drupal-project#updating-drupal-core
) for example 
I propose we have something like:
```
# Downloads cibox tagged with 1.0.6 in temporary folder and copies files onto your project repo.
./cibox-scaffold --version=1.0.6
# Merge any conflicts and push to your project repo.
```
I like what composer is doing and what it can do for our project, I don't mind using composer for non-PHP projects, because you can get some great benefits by using composer features like plugins and dependency management. At some point maybe we even split our playbooks into separate repos and have them versioned as well, composer will be able to help.

> - LTS versions should have a rethinked recipe for how to work with an enterprise projects like @propeoplemd/cypress-pp where we have a lot of servers for a production and getting builds is a pain. We have lack of tools for minimizing production dumps and optimizing reinstall process

We wanted to find a way to do database copies on files level and not doing a full dump, I think that would be a huuuge improvement. Platform.sh is doing database clones for environments in seconds for very large datasets just by using a file system approach, rather then dump -> import one. We should investigate more in that direction.

> LTS versions should have an ability to disable all reinstall steps by users. We should have features.yml config for that and good documentation.
The current state of it is you tweak your reinstall.yml as you like for your specific project needs, but the more it's configurable from some config file, the less users have to change in the actual reinstall.yml source. I think env vars and global vars folders already try to do that, but if we want to make some automatic updates for CIBOX we need to come up with a list of folders/paths that are considered project-specific and should not be updated during a CIBOX update.

**Comment by tylersamples** [comment]
I completely agree with the notion that CIBox should be packaged as a dependency, composer will work great, this has been suggested before in issue 331. 

However, I don't agree with a LTS version. LTS's are a broken model of development releases. It divides development resources between different versions, it encourages difficulty in patching security issues -- especially when major features land, and it encourages delaying major features in fixed release increments should CIBox ever move in that direction.

> At some point maybe we even split our playbooks into separate repos and have them versioned as well, composer will be able to help

I'd like to see this as well, in addition to this is clean abstractions.

> We wanted to find a way to do database copies on files level and not doing a full dump, I think that would be a huuuge improvement. Platform.sh is doing database clones for environments in seconds for very large datasets just by using a file system approach, rather then dump -> import one. We should investigate more in that direction.

An alternative I've been looking into is ro mounting of a production database and cloning on write for multi-dev environments by altering the network settings of a container; however this would require CIBox to use containers. Just a thought for future attempts at implementing this.
